### ISSUE TITLE
browserify always bundles the minified file

### ISSUE THREAD

**Issue by Prinzhorn** [regl-project/regl#464]
I'm trying to understand what is happening with my local build and I'm trying to figure out what this transform (#326) is supposed to do.

The issue I'm having is that regl is the only dependency in my build which is minified in development, which makes it hard/impossible to debug.

Steps to reproduce in a clean directory

```
npm i regl
echo "require('regl')" > index.js
npx browserify index.js > bundle.js
```

The bundle.js now contains the minified regl.

What is the purpose of the browserify transform that was added in #326? If I remove it, browserify correctly uses the `main` file specified in package.json and I get a non-minified bundle.js

**Comment by rreusser** [comment]
@Prinzhorn I believe you need to add the `--debug` flag to your browserify args. Then it will swap out regl with the unminified version. TBH I often just explicitly `require('regl/regl')` when I want to dig into node_modules and debug the internals.

You can test this by running:

```
echo "require('regl')" | browserify - --debug
```

which has the unminified source as opposed to

```
echo "require('regl')" | browserify
```

which uses the minified source.

**Comment by rreusser** [comment]
Since this is [noted in the documentation](https://github.com/regl-project/regl/blob/master/API.md#error-messages-and-debug-mode):

> By default if you compile regl with browserify then all error messages and checks are removed. This is done in order to reduce the size of the final bundle and to improve run time performance.
> 
> If you want error messages and are using browserify make sure that you compile using --debug. Not only will this insert debug messages but it will also give you source maps which make finding errors easier.
> 
> Alternatively, consider using budo for your live development server. budo automatically compiles your code in debug mode and also provides facilities for live reloading.

I'll close this issue, but would gladly accept a PR if there's an inconsistency I'm not catching or if there's something you think would make this more straightforward.

**Comment by Prinzhorn** [comment]
Thanks, I understand.

> I'll close this issue, but would gladly accept a PR if there's an inconsistency I'm not catching or if there's something you think would make this more straightforward.

As far as I understand, other projects (react and my projects included) use `NODE_ENV` for this purpose. When combined with uglify dead code elimination it behaves as you described.

Examples:

https://github.com/facebook/react/search?q=node_env&type=Code
https://github.com/reduxjs/redux/search?q=node_env&type=Code

I, as a user regl somewhere in my project, will then set `NODE_ENV` globally and get either a debug build (with ALL modules being unminified) or a production bundle (with ALL modules being compressed).

In dev `if(process.env.NODE_ENV === 'development')` is compiled to `if('development' === 'development')`. In production it turns into `if('production' === 'development')` and is removed by uglify.

[envify](https://www.npmjs.com/package/envify) does that, which is uncluded in https://www.npmjs.com/package/tinyify 

> envify - Replace environment variables—by default, replaces NODE_ENV with "production"

**Comment by rreusser** [comment]
I think the particular reason for the extra step is that there's some [rollup setup](https://github.com/regl-project/regl/blob/8f26f111a913cd9f7f239ebcdf691189e82c7a5f/rollup/plugins/remove-check.js) to remove checks at the AST level so that you don't need to write the conditional `if (process.env.NODE_ENV === 'development') check.command(...);` throughout the codebase in order for it to get removed in the production version. Alternatively you could put the conditional *inside* `check.command`, but then if you're reallly trying to squeeze out performance, you still have a large number of no-op function calls that don't get removed.

In other words, I believe it leans on rollup to perform the removal since it's more targeted and concise than uglify—so that the transform swaps the rollup'd version as needed.

I'm not sure it's a perfect solution (or that a perfect solution exists), but glad to help if it can be made clearer, whether through docs or tweaks that preserve the fundamental goals!

**Comment by Prinzhorn** [comment]
Now that I know about `--debug` i can use it instead of manually patching regl inside node_modules. However, I believe it is a misuse of the flag since it has side effects not intended by browserify

> When opts.debug is true, add a source map inline to the end of the bundle. This makes debugging easier because you can see all the original files if you are in a modern enough browser.

Or does it? As far as I understand you're shipping the unminified regl when --debug is set, instead of a source map?

> so that you don't need to write the conditional if (process.env.NODE_ENV === 'development') check.command(...); throughout the codebase

I see. I does make sense what you're saying, but it looks like a homebrown solution specific to this repository if I understand https://github.com/regl-project/regl/blob/8f26f111a913cd9f7f239ebcdf691189e82c7a5f/rollup/plugins/remove-check.js correctly.
